### PR TITLE
fix: fix grad accumulation/overlap under ZeRO-2

### PR DESCRIPTION
**Pull request by Chamberlain0w0**
## 背景

先前的实现 `is_last_microbatch` 的处理过于草率，导致目前 ZeRO-2 同时开启梯度累积和 `overlap_grad_reduce` 时，当前实现会在第一个 microbatch backward 期间发起 reduce-scatter，并将 `grad_reduce_dispatched_` 设置为 `true`。

该状态直到 `optimizer->step()` 调用 `FinishGradSync()` 后才会重置，因此存在两个问题：

1. 后续 microbatch 不会再次发起梯度规约，新增梯度无法参与本轮 reduce-scatter。
2. 第一个 microbatch 的异步 reduce-scatter 可能尚未结束，后续 microbatch 已开始写入 `temp_full_grad_buffer`，造成计算与通信之间的数据竞争。

关闭 `overlap_grad_reduce` 时不会触发该问题，因为梯度同步统一在所有 microbatch 完成后的 `optimizer->step()` 中执行。

## 修改内容

参考 Megatron-LM 的 `no_sync` 机制，引入真实的 `is_last_microbatch_` 控制：

- 为 `Module` 增加通用的 `no_sync()` 接口和 RAII `NoSyncGuard`。
- `DistributedDataParallel::no_sync()` 在 guard 生命周期内将 bucket group 的 `is_last_microbatch_` 设置为 `false`，退出时恢复为 `true`。
- 非最后一次 backward 仅将梯度累积到本地 buffer，不标记 bucket ready，也不发起 reduce-scatter。
- 最后一次 backward 恢复梯度同步，使完整的累积梯度只进行一次规约。
- 未开启 `overlap_grad_reduce` 时保持原有行为，由 `optimizer->step()` 发起同步。
- Pipeline Parallel 通过注入的 `no_sync_func_` 使用该机制，不直接依赖或包含 DDP 实现。
- PP/VPP 根据每个本地 model chunk 的实际 backward 顺序确定最后一次 backward，避免将 microbatch ID 错误地等同于 backward 完成顺序。
- GPT-2 和 LLaMA3 的非 PP 训练循环在非最后一个梯度累积 microbatch 中持有 `NoSyncGuard`。
- 不新增 `overlap_grad_reduce` 命令行参数，继续使用 DDP 配置中的默认行为。

## 行为变化

开启梯度累积和 `overlap_grad_reduce` 后，同一个 optimizer step 内的执行过程变为：

1. 非最后一个 microbatch：仅累积本地梯度。
2. 最后一个 microbatch：bucket ready 后发起一次异步 reduce-scatter。
3. `optimizer->step()`：等待通信完成并更新参数。

这样可以确保 reduce-scatter 读取的是所有 microbatch 累积后的完整梯度，同时避免通信期间继续修改 full gradient buffer。

### CODE REVIEW

File: infini_train/include/nn/modules/module.h
Comment (kilinchange) [inline infini_train/include/nn/modules/module.h:1]: 实现放到 .cc 里。

File: infini_train/include/nn/parallel/ddp/param_and_grad_buffer.h
Comment (kilinchange) [inline infini_train/include/nn/parallel/ddp/param_and_grad_buffer.h:1]: 实现放到 .cc 里。

File: infini_train/src/nn/parallel/ddp/param_and_grad_buffer.cc [hunk 1]
```diff
@@ -297,7 +301,6 @@ void ParamAndGradBucketGroup::StartGradSync() {
     }
 
     grad_reduce_dispatched_ = true;
-    // TODO(zbl): no need to clear params_with_grad_ here if grad sync is only done on last microbatch
```
Comment (kilinchange) [inline infini_train/src/nn/parallel/ddp/param_and_grad_buffer.cc:300]: 我看注释里说如果只在 last microbatch sync 的话，这里就不需要 clear 了，但下面的 clear 没有删，确认下是否需要删除？

File: infini_train/include/nn/parallel/pp/pipeline_schedule.h [hunk 2]
```diff
@@ -31,12 +33,17 @@ class PipelineSchedule {
                                    const std::vector<std::shared_ptr<Tensor>> &target_mbs,
                                    const std::shared_ptr<nn::Module> &loss_fn, DataType dtype);
 
+    using NoSyncFunc = std::function<std::vector<std::unique_ptr<nn::NoSyncGuard>>()>;
+
+    void SetNoSyncFunc(NoSyncFunc func) { no_sync_func_ = std::move(func); }
```
Comment (kilinchange) [inline infini_train/include/nn/parallel/pp/pipeline_schedule.h]: 实现放到 .cc 里

File: infini_train/include/nn/parallel/pp/pipeline_schedule.h [hunk 3]
```diff
@@ -1,9 +1,11 @@
 #pragma once
 
+#include <functional>
 #include <memory>
 #include <vector>
 
 #include "infini_train/include/datatype.h"
+#include "infini_train/include/nn/modules/module.h"
```
Comment (kilinchange) [inline infini_train/include/nn/parallel/pp/pipeline_schedule.h]: 前置声明 NoSyncGuard 就行，这里不需要 include module.h

File: example/llama3/main.cc [hunk 4]
```diff
@@ -465,6 +474,10 @@ void Train(const nn::parallel::Rank &rank) {
                 LOG(INFO) << "Rank " << rank.GlobalRank() << ": finish loss forward";
 
                 LOG(INFO) << "Rank " << rank.GlobalRank() << ": start backward";
+                std::unique_ptr<nn::NoSyncGuard> no_sync_guard;
+                if (ddp_world_size > 1 && FLAGS_zero_stage >= 1 && micro_step != grad_accum_steps - 1) {
```
Comment (kilinchange) [inline example/llama3/main.cc]: zero_stage=0 且开启梯度累积时，仍然会所有 microbatch 都进行梯度通信，但实际上也只需要最后一个 microbatch 发起通信，看看这个 pr 里方不方便一起改了？不好改的话就留个 FIXME 吧

**Review by kilinchange — CHANGES_REQUESTED** [review]

File: infini_train/src/nn/parallel/ddp/param_and_grad_buffer.cc (hunk 1 above)
Comment (Chamberlain0w0) [reply infini_train/src/nn/parallel/ddp/param_and_grad_buffer.cc:300]: 可以删，这块是忘删了。

File: example/llama3/main.cc (hunk 4 above)
Comment (Chamberlain0w0) [reply example/llama3/main.cc]: 统一解决了一下，现在三条 DDP 的路径都完成了在梯度累积下的 last microbatch 再发起梯度同步：
1. naive DDP（不带分桶）：给 DDP class 里加了一个 atomic_bool 的 is_last_microbatch_ 成员，用来标记是否为 last mb；`DDP::SetIsLastMicrobatch()` 会更新它；然后修改 `AllReducePostAccumulateHook`， 让其额外接受一个 bool 的参数，来决定是否要真的执行 allreduce。
```cpp
AllReducePostAccumulateHook(
    function::ReduceOpType reduce_op,
    const ProcessGroup *pg = nullptr,
    std::shared_ptr<const std::atomic_bool> enabled = nullptr);
```
2. 带分桶 DDP：给 Reducer class 里加了一个 bool 的 is_last_microbatch_ 成员，用来标记是否为 last mb；`DDP::SetIsLastMicrobatch()` 会更新它；BucketHook 入口 `Reducer::MarkVariableReadyDense` 处判断此 bool，来决定是否要继续执行标记 grad ready 开启通信的逻辑。

3. DistOpt-based ZeRO DDP（本 PR 原本核心改动）：给 BucketGroup class 里面加了一个 bool 的 is_last_microbatch_ 成员，用来标记是否为 last mb；`DDP::SetIsLastMicrobatch()` 会更新它；准备更新梯度的时候，在`ParamAndGradBucketGroup::RegisterGradReady` 入口进行判断，决定是否要进行梯度同步。

三条路径运行时互斥，实际运行时只会存在一种 DDP 的实现方式，所以 `DDP::SetIsLastMicrobatch()` 并不会产生冲突；在存在梯度累积时，三条 DDP 路径都通过本 PR 中核心修改的同一个 `NoSyncGuard` 来控制“非最后 microbatch 不通信”。

File: infini_train/include/nn/parallel/pp/pipeline_schedule.h (hunk 3 above)
Comment (Chamberlain0w0) [reply infini_train/include/nn/parallel/pp/pipeline_schedule.h]: 改了

File: infini_train/include/nn/parallel/pp/pipeline_schedule.h (hunk 2 above)
Comment (Chamberlain0w0) [reply infini_train/include/nn/parallel/pp/pipeline_schedule.h]: 改了

File: infini_train/include/nn/parallel/ddp/param_and_grad_buffer.h
Comment (Chamberlain0w0) [reply infini_train/include/nn/parallel/ddp/param_and_grad_buffer.h:1]: 改了

File: infini_train/include/nn/modules/module.h
Comment (Chamberlain0w0) [reply infini_train/include/nn/modules/module.h:1]: 改了

File: infini_train/src/nn/modules/module.cc [hunk 5]
```diff
@@ -21,10 +21,22 @@
...
+        exit_func_();
+    }
+}
+
 Module::Module() : Module(kUndefinedType) {}
 
 Module::Module(const std::string &type) : type_(type), device_(Device()) {}
 
+std::unique_ptr<NoSyncGuard> Module::no_sync() {
+    return std::make_unique<NoSyncGuard>([] {});
```
Comment (chen2021673) [inline infini_train/src/nn/modules/module.cc:37]: 这里return nullptr就好了吧

File: infini_train/src/nn/parallel/ddp/distributed_data_parallel.cc [hunk 6]
```diff
@@ -216,4 +216,17 @@ DistributedDataParallel::Forward(const std::vector<std::shared_ptr<Tensor>> &inp
 }
 
 std::shared_ptr<nn::Module> DistributedDataParallel::module() const { return modules_.at(kModuleName); }
+
+std::unique_ptr<nn::NoSyncGuard> DistributedDataParallel::no_sync() {
+    SetIsLastMicrobatch(false);
+    return std::make_unique<nn::NoSyncGuard>([this] { SetIsLastMicrobatch(true); });
```
Comment (chen2021673) [inline infini_train/src/nn/parallel/ddp/distributed_data_parallel.cc:222]: PyTorch 在这里和 Megatron 不太一样，进入 no_sync() 时保存旧状态，退出时会恢复旧状态，所以能支持嵌套作用域。目前我觉得不太需要，如果后续需要支持嵌套的话可以参考。

File: example/gpt2/main.cc [hunk 7]
```diff
@@ -290,11 +290,18 @@ void Train(const nn::parallel::Rank &rank) {
...
         if (ddp_world_size > 1) {
             auto ddp_config = DistributedDataParallelConfig{.zero_stage = FLAGS_zero_stage};
-            auto *mutable_chunks = dynamic_cast<nn::parallel::PipelineParallel *>(model.get())->mutable_chunks();
+            auto *pipeline_model = dynamic_cast<nn::parallel::PipelineParallel *>(model.get());
+            auto *mutable_chunks = pipeline_model->mutable_chunks();
             for (int chunk_id = 0; chunk_id < mutable_chunks->size(); ++chunk_id) {
                 (*mutable_chunks)[chunk_id]
                     = std::make_shared<DistributedDataParallel>(mutable_chunks->at(chunk_id), rank, ddp_config);
             }
+            pipeline_model->SetNoSyncFunc([mutable_chunks] {
```
Comment (chen2021673) [inline example/gpt2/main.cc:299]: 这里主要是给 Pipeline 每个 chunk 注册回调函数，但 PipelineSchedule  自己能访问 stage_->chunks，能不能让 PipelineSchedule  直接遍历 chunks 创建 guard 啊，这样相关的逻辑不用放在训练入口。

Comment (kilinchange) [reply example/gpt2/main.cc:299]: stage 里维护的 chunk 是 Module 基类类型，需要根据 ddp_world_size 判断，才能安全将 module 转换成 DistributedDataParallel 类型。因此不建议目前在 pp 模块内部引入 ddp 相关的耦合逻辑，后续训练入口这块可以考虑统一整理成类似 megatron train.py 的形式，提供一个统一的训练入口。

**Review by kilinchange — APPROVED** [review]

File: infini_train/src/nn/modules/module.cc (hunk 5 above)
Comment (Chamberlain0w0) [reply infini_train/src/nn/modules/module.cc]: 确实，我改下

**Review by chen2021673 — APPROVED** [review]

**Review by kilinchange — APPROVED** [review]
